In [129]:
import os
import IPython.display as ipd
import matplotlib.pyplot as plt

import numpy as np
import librosa, librosa.display
from scipy.fft import rfft
from sklearn.neighbors import KNeighborsClassifier
from sklearn.decomposition import PCA

#### The raw training/validation/test data are the paths to the audio files, labels for now are just the emotions. The gender and intensity lists maybe further used as features

In [5]:
raw_train_data = []
raw_val_data = []
raw_test_data = []

train_label = []
val_label = []
test_label = []

train_gender = []
val_gender = []
test_gender = []

train_intensity = []
val_intensity = []
test_intensity = []

for i in range(0, 16):
    # Get all file names in the dir
    actor = "Actor_%02d" % (i + 1)
    inputs = os.listdir('./data/RAV/' + actor)
    
    for ele in inputs:
        raw_train_data.append('./data/RAV/' + actor + '/' + ele) # Form paths to the files
        
        file_name = ele.split('-')
        train_label.append(int(file_name[2]))                    # Get emotion label
        train_intensity.append(int(file_name[3]))                # Get intensity
        train_gender.append((i + 1) % 2)                         # Get gender (1 for male, 0 for female)
    
for i in range(16, 20):
    # Get all file names in the dir
    actor = "Actor_%02d" % (i + 1)
    inputs = os.listdir('./data/RAV/' + actor)

    for ele in inputs:
        raw_val_data.append('./data/RAV/' + actor + '/' + ele) # Form paths to the files
        
        file_name = ele.split('-')
        val_label.append(int(file_name[2]))                    # Get emotion label
        val_intensity.append(int(file_name[3]))                # Get intensity
        val_gender.append((i + 1) % 2)                         # Get gender (1 for male, 0 for female)
        
for i in range(20, 24):
    # Get all file names in the dir
    actor = "Actor_%02d" % (i + 1)
    inputs = os.listdir('./data/RAV/' + actor)

    for ele in inputs:
        raw_test_data.append('./data/RAV/' + actor + '/' + ele) # Form paths to the files
        
        file_name = ele.split('-')
        test_label.append(int(file_name[2]))                    # Get emotion label
        test_intensity.append(int(file_name[3]))                # Get intensity
        test_gender.append((i + 1) % 2)                         # Get gender (1 for male, 0 for female)

#### Helper functions to split the data based on gender

In [152]:
def split_gender(train, train_label, train_gender, val, val_label, val_gender, test, test_label, test_gender):
    
    train_male, train_male_label, train_female, train_female_label = split_gender_helper(train, train_label, train_gender)
    val_male, val_male_label, val_female, val_female_label = split_gender_helper(val, val_label, val_gender)
    test_male, test_male_label, test_female, test_female_label = split_gender_helper(test, test_label, test_gender)


    return train_male, train_male_label, train_female, train_female_label, \
           val_male, val_male_label, val_female, val_female_label, \
           test_male, test_male_label, test_female, test_female_label

def split_gender_helper(data, label, gender):
    male = []
    male_label = []
    female = []
    female_label = []
    for i in range(0, len(data)):
        if gender[i] == 1:
            male.append(data[i])
            male_label.append(label[i])
        else:
            female.append(data[i])
            female_label.append(label[i])
    return male, male_label, female, female_label

#### Use librosa to load the audio files

In [47]:
lbr_train_data = []
for ele in raw_train_data:
    temp, _ = librosa.load(ele)
    lbr_train_data.append(np.pad(temp, (0, 116247 - len(temp))))

In [48]:
lbr_val_data = []
for ele in raw_val_data:
    temp, _ = librosa.load(ele)
    lbr_val_data.append(np.pad(temp, (0, 116247 - len(temp))))

In [49]:
lbr_test_data = []
for ele in raw_test_data:
    temp, _ = librosa.load(ele)
    lbr_test_data.append(np.pad(temp, (0, 116247 - len(temp))))

#### Get the Fourier transform of the data as features
The result of from the Fourier transform on the entire data set may not offer much information since all information about time in the sound is inherently lost. In human speech, having a high pitch at the beginning may indicate very different emotions from having a high pitch at the end. I do not expect great results from such features.

In [84]:
fft_train_data = []
for ele in lbr_train_data:
    temp = rfft(ele)
    fft_train_data.append(temp.real)
        
fft_val_data = []
for ele in lbr_val_data:
    temp = rfft(ele)
    fft_val_data.append(temp.real)
        
fft_test_data = []
for ele in lbr_test_data:
    temp = rfft(ele)
    fft_test_data.append(temp.real)

#### Using KNN classifier for FFT features
As seen here, FFT+KNN yields validation errors slightly better than chance, I would expect similar results for the test error, therefore, I did not evaluate based on the test data.

In [102]:
for n in [1,3,5,10,15,20,30,40,50]:
    knn_classifier = KNeighborsClassifier(n_neighbors = n)
    knn_classifier.fit(fft_train_data, train_label)
    print("n = " + str(n) + ":" + str(knn_classifier.score(fft_val_data, val_label)))

n = 1:0.11666666666666667
n = 3:0.11666666666666667
n = 5:0.125
n = 10:0.1
n = 15:0.125
n = 20:0.10833333333333334
n = 30:0.14166666666666666
n = 40:0.13333333333333333
n = 50:0.13333333333333333


Due to the high dimensionality of the FFT features, PCA is applied to reduce the effect of the Curse of Dimensionality and improve performance. From running the code, we can see that accuracy increased significantly (0.14 -> 0.24)

In [107]:
for nc in [1,2,4,8,16,32,64,128,256,512]:
    pca = PCA(n_components = nc)
    fft_train_pca = pca.fit_transform(fft_train_data)
    fft_val_pca = pca.transform(fft_val_data)
    fft_test_pca = pca.transform(fft_test_data)
    print("n_components = " + str(nc))
    
    for n in [1,3,5,10,15,20,30,40,50]:
        knn_classifier = KNeighborsClassifier(n_neighbors = n)
        knn_classifier.fit(fft_train_pca, train_label)
        print("    n = " + str(n) + ":" + str(knn_classifier.score(fft_val_pca, val_label)))

n_components = 1
    n = 1:0.1625
    n = 3:0.17083333333333334
    n = 5:0.2
    n = 10:0.21666666666666667
    n = 15:0.175
    n = 20:0.2
    n = 30:0.19166666666666668
    n = 40:0.19583333333333333
    n = 50:0.19583333333333333
n_components = 2
    n = 1:0.15416666666666667
    n = 3:0.15833333333333333
    n = 5:0.19166666666666668
    n = 10:0.21666666666666667
    n = 15:0.2125
    n = 20:0.2125
    n = 30:0.19583333333333333
    n = 40:0.2
    n = 50:0.19166666666666668
n_components = 4
    n = 1:0.2
    n = 3:0.16666666666666666
    n = 5:0.2125
    n = 10:0.23333333333333334
    n = 15:0.225
    n = 20:0.23333333333333334
    n = 30:0.22916666666666666
    n = 40:0.2375
    n = 50:0.2375
n_components = 8
    n = 1:0.2375
    n = 3:0.2
    n = 5:0.20416666666666666
    n = 10:0.19583333333333333
    n = 15:0.21666666666666667
    n = 20:0.17916666666666667
    n = 30:0.21666666666666667
    n = 40:0.2375
    n = 50:0.23333333333333334
n_components = 16
    n = 1:0.1625
    n

In [165]:
# Find the test error for a good setting of hyper-parameters.
pca = PCA(n_components = 4)
fft_train_pca = pca.fit_transform(fft_train_data)
fft_val_pca = pca.transform(fft_val_data)
fft_test_pca = pca.transform(fft_test_data)

knn_classifier = KNeighborsClassifier(n_neighbors = 50)
knn_classifier.fit(fft_train_pca, train_label)

knn_classifier.score(fft_test_pca, test_label)

0.2

In [121]:
pca = PCA(n_components = 4)
fft_train_pca = pca.fit_transform(fft_train_data)
sum(pca.explained_variance_ratio_)

0.10590674833007543

### Splitting the data based on gender (with PCA)

In [177]:
for nc in [1,2,4,8,16,32,64,128,256,512]:
    pca = PCA(n_components = nc)
    fft_train_pca = pca.fit_transform(fft_train_data)
    fft_val_pca = pca.transform(fft_val_data)
    fft_test_pca = pca.transform(fft_test_data)
    print("n_components = " + str(nc))

    fft_train_male, train_male_label, fft_train_female, train_female_label, \
    fft_val_male, val_male_label, fft_val_female, val_female_label, \
    fft_test_male, test_male_label, fft_test_female, test_female_label = \
        split_gender(fft_train_pca, train_label, train_gender, \
                     fft_val_pca, val_label, val_gender, \
                     fft_test_pca, test_label, test_gender)
        
        
    print("---------male-----------")
    for n in [1,3,5,10,15,20,30,40,50]:
        knn_male = KNeighborsClassifier(n_neighbors = n)
        knn_male.fit(fft_train_male, train_male_label)
        print("    n = " + str(n) + ":" + str(knn_male.score(fft_val_male, val_male_label)))
        
    print("--------female----------")
    for n in [1,3,5,10,15,20,30,40,50]:
        knn_female = KNeighborsClassifier(n_neighbors = n)
        knn_female.fit(fft_train_female, train_female_label)
        print("    n = " + str(n) + ":" + str(knn_female.score(fft_val_female, val_female_label)))

n_components = 1
---------male-----------
    n = 1:0.16666666666666666
    n = 3:0.18333333333333332
    n = 5:0.15
    n = 10:0.16666666666666666
    n = 15:0.225
    n = 20:0.225
    n = 30:0.23333333333333334
    n = 40:0.225
    n = 50:0.21666666666666667
--------female----------
    n = 1:0.19166666666666668
    n = 3:0.15833333333333333
    n = 5:0.15833333333333333
    n = 10:0.16666666666666666
    n = 15:0.18333333333333332
    n = 20:0.19166666666666668
    n = 30:0.18333333333333332
    n = 40:0.15833333333333333
    n = 50:0.225
n_components = 2
---------male-----------
    n = 1:0.14166666666666666
    n = 3:0.175
    n = 5:0.20833333333333334
    n = 10:0.24166666666666667
    n = 15:0.225
    n = 20:0.2
    n = 30:0.225
    n = 40:0.26666666666666666
    n = 50:0.23333333333333334
--------female----------
    n = 1:0.175
    n = 3:0.13333333333333333
    n = 5:0.23333333333333334
    n = 10:0.15833333333333333
    n = 15:0.15
    n = 20:0.18333333333333332
    n = 30:0.

In [198]:
pca = PCA(n_components = 4)
fft_train_pca = pca.fit_transform(fft_train_data)
fft_val_pca = pca.transform(fft_val_data)
fft_test_pca = pca.transform(fft_test_data)

fft_train_male, train_male_label, fft_train_female, train_female_label, \
fft_val_male, val_male_label, fft_val_female, val_female_label, \
fft_test_male, test_male_label, fft_test_female, test_female_label = \
    split_gender(fft_train_pca, train_label, train_gender, \
                 fft_val_pca, val_label, val_gender, \
                 fft_test_pca, test_label, test_gender)
    
knn_male = KNeighborsClassifier(n_neighbors = 30)
knn_male.fit(fft_train_male, train_male_label)
knn_male.score(fft_test_male, test_male_label)

0.25

In [197]:
pca = PCA(n_components = 4)
fft_train_pca = pca.fit_transform(fft_train_data)
fft_val_pca = pca.transform(fft_val_data)
fft_test_pca = pca.transform(fft_test_data)

fft_train_male, train_male_label, fft_train_female, train_female_label, \
fft_val_male, val_male_label, fft_val_female, val_female_label, \
fft_test_male, test_male_label, fft_test_female, test_female_label = \
    split_gender(fft_train_pca, train_label, train_gender, \
                 fft_val_pca, val_label, val_gender, \
                 fft_test_pca, test_label, test_gender)
    
knn_female = KNeighborsClassifier(n_neighbors = 40)
knn_female.fit(fft_train_female, train_female_label)
knn_female.score(fft_test_female, test_female_label)

0.23333333333333334

## Good setting for KNN using FFT features with PCA: K = 30 / 40 / 50, n_components = 4 / 8. Validation accuracy = 0.2375, test accuracy = 0.225  
## Also, from experimenting, we can see that spliting the input by gender offers some improvements in performance, since similar behavior is expected for spectrogram features, such testing is not perfored on the stft data set

#### Get spectrograms of the audio input as features as to preserve the time dimension of the input

In [73]:
stft_train_data = []
for ele in lbr_train_data:
    temp = librosa.stft(ele, n_fft = 512)
    stft_train_data.append(temp.real.reshape(-1))
        
stft_val_data = []
for ele in lbr_val_data:
    temp = librosa.stft(ele, n_fft = 512)
    stft_val_data.append(temp.real.reshape(-1))
        
stft_test_data = []
for ele in lbr_test_data:
    temp = librosa.stft(ele, n_fft = 512)
    stft_test_data.append(temp.real.reshape(-1))

#### Using KNN with spectrogram as features  
Due to the high number of variables in a spectrogram, making a prediction with this model is extremely slow. The validation error seems to be slightly better than using FFT features, perhaps I would need to shift the spectrograms in the training data set to get better performance.

In [74]:
# Training the knn classifier
knn_classifier = KNeighborsClassifier(n_neighbors = 30)
knn_classifier.fit(stft_train_data, train_label)

KNeighborsClassifier(n_neighbors=30)

In [81]:
for n in [1,3,5,10,15,20,30,40,50]:
    knn_classifier = KNeighborsClassifier(n_neighbors = n)
    knn_classifier.fit(stft_train_data, train_label)
    print("n = " + str(n) + ":" + str(knn_classifier.score(stft_val_data, val_label)))

1:0.1375
3:0.11666666666666667
5:0.12083333333333333
10:0.10416666666666667
15:0.125
20:0.10833333333333334
30:0.14166666666666666
40:0.13333333333333333
50:0.13333333333333333


In [112]:
for nc in [1,2,4,8,16,32,64,128,256,512]:
    pca = PCA(n_components = nc)
    stft_train_pca = pca.fit_transform(stft_train_data)
    stft_val_pca = pca.transform(stft_val_data)
    stft_test_pca = pca.transform(stft_test_data)
    print("n_components = " + str(nc))
    
    for n in [1,3,5,10,15,20,30,40,50]:
        knn_classifier = KNeighborsClassifier(n_neighbors = n)
        knn_classifier.fit(stft_train_pca, train_label)
        print("    n = " + str(n) + ":" + str(knn_classifier.score(stft_val_pca, val_label)))

n_components = 1
    n = 1:0.16666666666666666
    n = 3:0.16666666666666666
    n = 5:0.16666666666666666
    n = 10:0.20833333333333334
    n = 15:0.22083333333333333
    n = 20:0.225
    n = 30:0.19166666666666668
    n = 40:0.20416666666666666
    n = 50:0.225
n_components = 2
    n = 1:0.175
    n = 3:0.17916666666666667
    n = 5:0.19583333333333333
    n = 10:0.2125
    n = 15:0.23333333333333334
    n = 20:0.2125
    n = 30:0.22916666666666666
    n = 40:0.21666666666666667
    n = 50:0.20416666666666666
n_components = 4
    n = 1:0.2125
    n = 3:0.19166666666666668
    n = 5:0.19583333333333333
    n = 10:0.19166666666666668
    n = 15:0.22916666666666666
    n = 20:0.225
    n = 30:0.225
    n = 40:0.2375
    n = 50:0.2375
n_components = 8
    n = 1:0.17083333333333334
    n = 3:0.1625
    n = 5:0.22083333333333333
    n = 10:0.20416666666666666
    n = 15:0.225
    n = 20:0.2125
    n = 30:0.22916666666666666
    n = 40:0.23333333333333334
    n = 50:0.23333333333333334
n_c

In [114]:
# Find the test error for a good setting of hyper-parameters.
pca = PCA(n_components = 4)
stft_train_pca = pca.fit_transform(stft_train_data)
stft_val_pca = pca.transform(stft_val_data)
stft_test_pca = pca.transform(stft_test_data)

knn_classifier = KNeighborsClassifier(n_neighbors = 40)
knn_classifier.fit(stft_train_pca, train_label)

knn_classifier.score(stft_test_pca, test_label)

0.24583333333333332

In [120]:
pca = PCA(n_components = 4)
stft_train_pca = pca.fit_transform(stft_train_data)
sum(pca.explained_variance_ratio_)

0.10513143032743553

## Good setting of hyper-parameters for Spectrogram features with KNN and PCA for dimension reduction: K = 40 / 50, n_components = 4. Validation accuracy = 0.2375, test accuracy = 0.24583

### From the low explained_variance_ratio_ of the trained PCA for both FFT and STFT (Spectrogram) features, I believe that FFT and Spectrogram features are not suited for the KNN algorithm for the following reasons:  
 - Low explained variance ratio implies that PCA doesn't do a very good job at reducing the dimensionality of the input
 - KNN has bad performance with data that has high dimensionality
 - Thus a balance between the dimensionality of the feature space and explained variance ratio after using PCA that results in good performance cannot be achieved